controllers/etcd_custodian_controller.go

```diff
@@ -80,12 +125,183 @@ func (ec *EtcdCustodian) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.
+
+		renewTime := metav1.NewMicroTime(time.Now())
+		deltaLease = &coordinationv1.Lease{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "delta-snapshot-revisions",
+				Namespace: etcd.Namespace,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion:         "druid.gardener.cloud/v1alpha1",
+						BlockOwnerDeletion: pointer.BoolPtr(true),
+						Controller:         pointer.BoolPtr(true),
+						Kind:               "Etcd",
+						Name:               etcd.Name,
+						UID:                etcd.UID,
+					},
+				},
+			},
+			Spec: coordinationv1.LeaseSpec{
+				HolderIdentity: pointer.StringPtr("0"),
+				RenewTime:      &renewTime,
+			},
+		}
+		err2 := ec.Create(ctx, deltaLease)
+		if err2 != nil {
+			logger.Error(err2, "Delta snap lease delta-snapshot-revisions couldn't be created")
+			return ctrl.Result{
+				RequeueAfter: 5 * time.Second,
+			}, nil
+		}
+
+		deltaLease = &coordinationv1.Lease{}
+		err2 = ec.Get(ctx, nsName, deltaLease)
+		if err2 != nil {
+			logger.Error(err2, "Delta snap lease delta-snapshot-revisions couldn't be fetched just after creation")
+			return ctrl.Result{
+				RequeueAfter: 5 * time.Second,
+			}, nil
+		}
+	}
+
+	if !checkEtcdOwnerReference(deltaLease.GetOwnerReferences(), etcd) {
+		err := kutil.TryPatch(ctx, retry.DefaultBackoff, ec.Client, deltaLease, func() error {
+			deltaLease.OwnerReferences = []metav1.OwnerReference{
+				{
+					APIVersion:         "druid.gardener.cloud/v1alpha1",
+					BlockOwnerDeletion: pointer.BoolPtr(true),
+					Controller:         pointer.BoolPtr(true),
+					Kind:               "Etcd",
+					Name:               etcd.Name,
+					UID:                etcd.UID,
+				},
+			}
+			return nil
+		})
+		if err != nil {
+			logger.Error(err, "Delta snap lease found but error occured during claiming it")
+			return ctrl.Result{
+				RequeueAfter: 5 * time.Second,
+			}, nil
+		}
+	}
+	// Run compaction job
+	if etcd.Spec.Backup.Store != nil {
+		hi, err := strconv.ParseInt(*deltaLease.Spec.HolderIdentity, 10, 32)
+		if err != nil {
+			logger.Error(err, "Can't convert holder identity of revision lease to integer")
+			return ctrl.Result{}, err
+		}
+		if hi >= ec.config.EventsThreshold {
+			// First check if a job is already running
+			// Reconcile job only when current revision is more than 1 million (this vale is configurable through `events-threshold` option in druid) than
+			// the last snapshot (full/compact) revision
+			values, err := getMapFromEtcd(ec.ImageVector, etcd)
+			if err != nil {
+				logger.Error(err, "Error while extracting values from ETCD spec")
+				return ctrl.Result{}, err
+			}
+
+			chartPath := getChartPath()
+			renderedChart, err := ec.chartApplier.Render(chartPath, etcd.Name, etcd.Namespace, values)
+			if err != nil {
+				logger.Error(err, "Error while rendering chart")
+				return ctrl.Result{}, err
+			}
+
+			j, err := ec.reconcileJob(ctx, logger, etcd, renderedChart)
```

Comment:
I suggest to add a new reconciler/controller which is solely responsible for creating/reconciling the job. It's a clear separation of concern and not really related to what the custodian does today (mainly health checks). This controller can then watch the `Leases` and create the jobs whenever necessary.

Reply:
Okay.. Makes sense.

Reply:
@timuthy so a new controller like `compaction_controller`? But if I remember right, we wanted `custodian_controller` to be taking care of etcd operations that had to be done outside the shoot maintenance window, so compaction would still fit into the responsibilities of `custodian_controller`.

Reply:
> we wanted custodian_controller to be taking care of etcd operations

Hm, I think we should restrict that responsibility to the etcd cluster only. Compaction isn't really related to the cluster directly, IMHO. It's rather a separate feature with a separate reconcile trigger. So why not clearly isolating it with a dedicated controller instead of squeezing it into the Custodian and worsen code maintenance in the future?